### ISSUE TITLE
Support for OSC 9;9

### ISSUE THREAD

**Issue by floh96** [JanDeDobbeleer/oh-my-posh#409]
It would be nice if the oh-my-posh v3 powershell module would support OSC 9;9. It was supported in [v2](https://github.com/JanDeDobbeleer/oh-my-posh/blob/master/oh-my-posh.psm1#L38).
Windows Terminal added support for OSC 9;9 for the preview version (1.6) recently see https://github.com/microsoft/terminal/pull/8330 and https://github.com/microsoft/terminal/pull/8934.

**Comment by JanDeDobbeleer** [comment]
That's an easy addition. Can do.

**Comment by floh96** [comment]
@JanDeDobbeleer do i need to configure it? it's not working in powershell 7 on windows 10 in windows terminal preview (1.6).
(without oh-my-posh i added [this](https://gist.github.com/skyline75489/480d036db8ae9069b7009377e6eebb79) to my profile and it's working fine.)

**Comment by JanDeDobbeleer** [comment]
@floh96 you updated right? I didn't validate it yet, but that's exactly what's added to the prompt now.

**Comment by floh96** [comment]
yeah i'm using version 3.89.0

**Comment by JanDeDobbeleer** [comment]
I'll check this evening maybe I did something stupid

**Comment by tradiff** [comment]
Alright I'll go ahead and ask. I skimmed the linked issues, and the PR Jan wrote, and I don't get it.
ELI5, what's OSC 9;9?

**Comment by floh96** [comment]
@TravisTX it's an ansi escape sequence that informs the terminal (windows terminal and conemu are supporting it) about the current working directory. When your current working directory is _C:\Users_ for example and you are duplicating your tab in windows terminal with ctrl+shift+d (the keybinding is not bound in the default configuration, you must manually add it to your settings.json in windows terminal) the working directory of the new tab is also _C:\Users_.

**Comment by tradiff** [comment]
I built OMP3 from source, the way I usually do, and can confirm this isn't working for me either.  I confirmed OMP3 is emitting something that looks correct with my limited knowledge of this. Here's the output, base64 encoded:
```
G105Ozk7QzpcUHJvamVjdHNcUGVyc29uYWxcb2gtbXktcG9zaDNcc3JjGzc=
```

CTRL+SHIFT+D opens a new tab with my normal starting directory.
Windows Terminal Version: 1.5.10411.0
PowerShell 7.1.2

**Comment by floh96** [comment]
@TravisTX it's only supported in the windows terminal preview version (1.6) atm i should have mentioned that

**Comment by tradiff** [comment]
Test preview also. Same result.
Windows Terminal Preview Version: 1.6.10412.0

**Comment by JanDeDobbeleer** [comment]
Maybe the quotes are required.

**Comment by floh96** [comment]
~~yeah i think the quotes are required see https://github.com/microsoft/terminal/pull/8934#issuecomment-769587247~~ Edit: I don't think the quotes are the problem see https://github.com/skyline75489/terminal/blob/66ce9e2917aa2fb81cf3bbf6613f3a15e643b87e/src/cascadia/TerminalCore/TerminalDispatch.cpp#L486

**Comment by JanDeDobbeleer** [comment]
@TravisTX @floh96 I adjusted the sequence according to [what's expected](https://github.com/microsoft/terminal/pull/8934/files#diff-7633954314d06182fca274ce5eab97b3d1c6d318f2023a6aa02ceb19912e33ddR416) and yet still it doesn't work for V3. A bit puzzled here tbh.

**Comment by lnu** [comment]
But is it integrated in the latest windows terminal preview? 
Even if I try manually with printf it does not work.
something like this should work ```printf "\e]9;9;C:\\\x9c"```

**Comment by JanDeDobbeleer** [comment]
@lnu fair point, I assumed so. Allow me to validate on ConEmu.

**Comment by floh96** [comment]
@lnu it is working with windows terminal preview 1.6 without oh-my-posh for me if i add this [gist](https://gist.github.com/skyline75489/480d036db8ae9069b7009377e6eebb79) to my powershell profile
also OSC 9;9 is only working on windows paths see [comment](https://github.com/microsoft/terminal/issues/8166#issuecomment-724132906).
Also you can see in the linked pull requests that it's released in the 1.6 preview.

**Comment by lnu** [comment]
@JanDeDobbeleer maybe it has something to do with the config of your windows terminal like default starting folder or something like that

**Comment by JanDeDobbeleer** [comment]
@lnu I feel a bit like an idiot as it should work in my latest PR according to the spec. I'll take a look tomorrow during my lunch break, maybe I made a silly mistake I just don't see now.

**Comment by chenxiaolong** [comment]
Could it be because of the ending byte sequence? With oh-my-posh 2, it ends with `\x1b\x37`:
```
0000000000000470 5B 4B 1B 5B 30 6D 20 1B 5B 4B 1B 5D 39 3B 39 3B [K�[0m �[K�]9;9;
0000000000000480 43 3A 5C 55 73 65 72 73 5C 63 68 65 6E 78 69 61 C:\Users\chenxia
0000000000000490 6F 6C 6F 6E 67 5C 44 6F 77 6E 6C 6F 61 64 73 1B olong\Downloads�
00000000000004A0 37                                              7
```
WIth oh-my-posh3 on the `main` branch, it ends with `\x07`:
```
0000000000000020 73 07 1B 5D 39 3B 39 3B 22 43 3A 5C 55 73 65 72 s��]9;9;"C:\User
0000000000000030 73 5C 63 68 65 6E 78 69 61 6F 6C 6F 6E 67 5C 44 s\chenxiaolong\D
0000000000000040 6F 77 6E 6C 6F 61 64 73 22 07                   ownloads"�
```
With oh-my-posh3 on the `fix-path-notification` branch, it ends with `\xc2\xa3`:
```
0000000000000470 5B 4B 1B 5B 30 6D 20 1B 5B 4B 1B 5D 39 3B 39 3B [K�[0m �[K�]9;9;
0000000000000480 22 43 3A 5C 55 73 65 72 73 5C 63 68 65 6E 78 69 "C:\Users\chenxi
0000000000000490 61 6F 6C 6F 6E 67 5C 44 6F 77 6E 6C 6F 61 64 73 aolong\Downloads
00000000000004A0 22 C2 A3                                        "Â£
```

**Comment by chenxiaolong** [comment]
I got it working on the `fix-path-notification` branch with either of these additional changes:
```diff
diff --git a/src/ansi_formats.go b/src/ansi_formats.go
index 0306d09..fb0e910 100644
--- a/src/ansi_formats.go
+++ b/src/ansi_formats.go
@@ -76,7 +76,7 @@ func (a *ansiFormats) init(shell string) {
 		a.escapeLeft = ""
 		a.escapeRight = ""
 		a.hyperlink = "\x1b]8;;%s\x1b\\%s\x1b]8;;\x1b\\"
-		a.osc99 = "\x1b]9;9;\"%s\"\x9c"
+		a.osc99 = "\x1b]9;9;\"%s\"\x1b\\"
 	}
 }
 
```
or
```diff
diff --git a/src/ansi_formats.go b/src/ansi_formats.go
index 0306d09..778251e 100644
--- a/src/ansi_formats.go
+++ b/src/ansi_formats.go
@@ -76,7 +76,7 @@ func (a *ansiFormats) init(shell string) {
 		a.escapeLeft = ""
 		a.escapeRight = ""
 		a.hyperlink = "\x1b]8;;%s\x1b\\%s\x1b]8;;\x1b\\"
-		a.osc99 = "\x1b]9;9;\"%s\"\x9c"
+		a.osc99 = "\x1b]9;9;\"%s\"\x07"
 	}
 }
 
```

I was looking at https://conemu.github.io/en/AnsiEscapeCodes.html#ConEmu_specific_OSC:

> Note. These codes may ends with ‘ESC\’ (two symbols - ESC and BackSlash) or ‘BELL’ (symbol with code \x07, same as ‘^a’ in *nix). For simplifying, endings in the following table marked as ‘ST’.

My understanding is that the sequence can either end in `\x1b\\` (ie. `\x1b\x5c`) or `\x07` by itself.

**Comment by lnu** [comment]
You're right. With this it works ```printf "\e]9;9;d:\\dev\e\\"``` which translates to ```\x1b]9;9d:\\dev\x1b\\``` 
~~But somehow, it works when duplicating a tab but not when splitting.~~

**Comment by JanDeDobbeleer** [comment]
I'll adjust, validate and merge! Thanks @chenxiaolong you made my day.

**Comment by lnu** [comment]
by the way, I've seen you named the template osc99. Maybe we should rename the hyperlink template to osc8.

**Comment by lnu** [comment]
Did anyone try with wsl. It crashes when duplicating on my side. I found this script while googling:https://gist.github.com/skyline75489/d655aede4c729eff178a1c0bfd10f622

**Comment by JanDeDobbeleer** [comment]
@lnu did you go for the first or second sample @chenxiaolong provided? The link you shared is using the first variation.

**Comment by lnu** [comment]
The same as the hyperlink. Ending with `\x1b\\`. The link comes from https://github.com/microsoft/terminal/issues/8166.

**Comment by lnu** [comment]
I also did some tests with the status indicator code osc 9;4 (https://github.com/microsoft/terminal/pull/8055). Would be cool to see the taskbar flashing when a long operation is running.

**Comment by lnu** [comment]
I confirm I'm not able to duplicate a  wsl tab with the osc9;9 code emitted. Without it, the tab is duplicated correctly.
![image](https://user-images.githubusercontent.com/1829553/107913362-94449c00-6f60-11eb-8847-9e898f8193a7.png)

To make it work with wsl, we have to do the same as [here](https://gist.github.com/skyline75489/d655aede4c729eff178a1c0bfd10f622):
```
_win_path=$(wslpath -m $(pwd))
printf "\033]9;9;%s\033\\" "$_win_path"
```

this code will correctly set the curent folder for wsl.

**Comment by JanDeDobbeleer** [comment]
@lnu this still doesn't work for me in WT (have the preview), can you validate my branch so we can merge?

**Comment by lnu** [comment]
let me check this. In the meantime I'm adding a function on environment to detect wsl or not and transform the path accordingly.

=> https://github.com/JanDeDobbeleer/oh-my-posh3/pull/413

**Comment by lnu** [comment]
@JanDeDobbeleer  it works with your banch on my side. Just one question: why using x033 instead of x1b like for hyperlinks? Both works so maybe we can just keep the same for all osc sequences?
I replace all 033 by x1b (also for title) and everything works.

**Comment by JanDeDobbeleer** [comment]
Yes, I was under the impression you said it didn't work with that. As what I see is @chenxiaolong first sample now?

**Comment by lnu** [comment]
@JanDeDobbeleer  Ok 033, x1b all the same :). so choose which one you want and you can use it for console, hyperlink and current folder. I tested both and it's ok. Sorry for the misunderstanding.

**Comment by JanDeDobbeleer** [comment]
I'm adjusting it

**Comment by JanDeDobbeleer** [comment]
@lnu swapped everything, if you an do a final I'll feel at ease here ;-)

**Comment by JanDeDobbeleer** [comment]
> let me check this. In the meantime I'm adding a function on environment to detect wsl or not and transform the path accordingly.

@lnu missed this one, in favour of merging your branch instead. Ping me on Twitter when done, I'm starting my day ;-)

**Comment by JanDeDobbeleer** [comment]
Available in [3.39.1](https://github.com/JanDeDobbeleer/oh-my-posh3/releases/tag/v3.89.1)

**Comment by lnu** [comment]
@floh96  Can you confirm all is working now with version 3.89.1?

**Comment by floh96** [comment]
@lnu i will check after work

**Comment by floh96** [comment]
@lnu @JanDeDobbeleer it's working now in powershell and wsl thank you ❤️

**Comment by JanDeDobbeleer** [comment]
@floh96 heads-up, I created a config switch for this at root level for the next version as enabling it by default effectively triggers a notification on iTerm2 😂

**Comment by lnu** [comment]
Sorry... just checked the doc of iterm2. The pandora box is opened: osc 9;9 not being standard across terminals, everyday will be a source of happiness 😅

**Comment by nova77** [comment]
Hey folks, this no longer seems to be working. I wonder if it's because instead of `$(wslpath -m $(pwd))` it should now be just `$PWD`.

**Comment by lnu** [comment]
Still working in pwsh but no more with wsl, checking why.

same as here https://github.com/JanDeDobbeleer/oh-my-posh/discussions/1508?

**Comment by lnu** [comment]
fixed here with https://github.com/JanDeDobbeleer/oh-my-posh/pull/1516

**Comment by joaociocca** [comment]
Is this what's causing my bash prompt to show this weird piece of code? (I've been searching everywhere for "9;9" and all I could find were binary matches in files, and this issue)

![image](https://user-images.githubusercontent.com/6473725/166509760-eff36cc7-9c3c-4c3a-aaa4-7fedaa3f5b5d.png)

also, if I <kbd>up</kbd> then <kbd>down</kbd> (go to the previous command in history then back to clear line), it goes away.

oh yeah, I'm using `sonicboom_light.omp.json` theme

**Comment by JanDeDobbeleer** [comment]
@joaociocca yes. And depending on the terminal you might need to disable that in the config (theme) as bot all of them support it.

**Comment by joaociocca** [comment]
@JanDeDobbeleer pardon my noobness, how can I disable it in this case?

**Comment by JanDeDobbeleer** [comment]
@joaociocca is that theme file on your file system? As there's a parameter in there `osc99` which is set to true. Only needs to be removed or set to false.

**Comment by joaociocca** [comment]
@JanDeDobbeleer yes, I downloaded a copy of `https://github.com/JanDeDobbeleer/oh-my-posh/blob/main/themes/sonicboom_light.omp.json`. Found the parameter, set to false and it's all pretty now! Thanks =D
